File: app/javascript/dashboard/helper/encryption.js
// Browser-compatible encryption utility using Web Crypto API
// This mirrors the Node.js crypto implementation for compatibility

const ENCRYPTION_KEY =
  window.chatwootConfig?.encryptionKey ||
  'default-encryption-key-change-in-production';
const ALGORITHM = 'AES-GCM';
const IV_LENGTH = 16;
const SALT_LENGTH = 32;
const TAG_LENGTH = 16;
const KEY_LENGTH = 32;

// Convert string to Uint8Array
function stringToUint8Array(str) {
  return new TextEncoder().encode(str);
}

// Convert Uint8Array to base64
function uint8ArrayToBase64(array) {
  return btoa(String.fromCharCode(...array));
}

// Convert base64 to Uint8Array
function base64ToUint8Array(base64) {
  const binary = atob(base64);
  const array = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i += 1) {
    array[i] = binary.charCodeAt(i);
  }
  return array;
}

// Derive key using PBKDF2 (similar to scrypt in Node.js)
async function deriveKey(salt) {
  const keyMaterial = await window.crypto.subtle.importKey(
    'raw',
    stringToUint8Array(ENCRYPTION_KEY),
    { name: 'PBKDF2' },
    false,
    ['deriveBits', 'deriveKey']
  );

  return window.crypto.subtle.deriveKey(
    {
      name: 'PBKDF2',
      salt,
      iterations: 100000, // Higher iterations for security
      hash: 'SHA-256',
    },
    keyMaterial,
    { name: ALGORITHM, length: KEY_LENGTH * 8 },
    false,
    ['encrypt', 'decrypt']
  );
}

/**
 * Encrypts a text string using AES-GCM
 * @param {string} text - The text to encrypt
 * @returns {Promise<string>} Base64 encoded encrypted data
 */
export async function encrypt(text) {
  // Generate random salt and IV
  const salt = window.crypto.getRandomValues(new Uint8Array(SALT_LENGTH));
  const iv = window.crypto.getRandomValues(new Uint8Array(IV_LENGTH));

  // Derive key from password and salt
  const key = await deriveKey(salt);

  // Encrypt the text
  const encodedText = stringToUint8Array(text);
  const encryptedBuffer = await window.crypto.subtle.encrypt(
    {
      name: ALGORITHM,
      iv,
    },
    key,
    encodedText
  );

  const encrypted = new Uint8Array(encryptedBuffer);

  // In AES-GCM, the authentication tag is included in the encrypted output
  // We need to extract it for compatibility with Node.js format
  const ciphertext = encrypted.slice(0, encrypted.length - TAG_LENGTH);
  const tag = encrypted.slice(encrypted.length - TAG_LENGTH);

  // Concatenate: salt + iv + tag + ciphertext
  const result = new Uint8Array(
    SALT_LENGTH + IV_LENGTH + TAG_LENGTH + ciphertext.length
  );
  result.set(salt, 0);
  result.set(iv, SALT_LENGTH);
  result.set(tag, SALT_LENGTH + IV_LENGTH);
  result.set(ciphertext, SALT_LENGTH + IV_LENGTH + TAG_LENGTH);

  // Return base64 encoded result
  return uint8ArrayToBase64(result);
}

/**
 * Decrypts a base64 encoded encrypted string
 * @param {string} encryptedText - Base64 encoded encrypted data
 * @returns {Promise<string>} Decrypted text
 */
export async function decrypt(encryptedText) {
  const buffer = base64ToUint8Array(encryptedText);

  // Extract components
  const salt = buffer.slice(0, SALT_LENGTH);
  const iv = buffer.slice(SALT_LENGTH, SALT_LENGTH + IV_LENGTH);
  const tag = buffer.slice(
    SALT_LENGTH + IV_LENGTH,
    SALT_LENGTH + IV_LENGTH + TAG_LENGTH
  );
  const ciphertext = buffer.slice(SALT_LENGTH + IV_LENGTH + TAG_LENGTH);

  // Derive key from password and salt
  const key = await deriveKey(salt);

  // Reconstruct the encrypted data with tag for AES-GCM
  const encryptedData = new Uint8Array(ciphertext.length + tag.length);
  encryptedData.set(ciphertext);
  encryptedData.set(tag, ciphertext.length);

  // Decrypt
  const decryptedBuffer = await window.crypto.subtle.decrypt(
    {
      name: ALGORITHM,
      iv,
    },
    key,
    encryptedData
  );

  // Convert back to string
  return new TextDecoder().decode(decryptedBuffer);
}

File: app/services/chatwoot_extra/encryption.rb
require 'openssl'
require 'base64'
require 'securerandom'

module ChatwootExtra
  module Encryption
    ALGORITHM = 'aes-256-gcm'.freeze
    IV_LENGTH = 16
    SALT_LENGTH = 32
    TAG_LENGTH = 16
    KEY_LENGTH = 32
    ITERATIONS = 100_000
    DEFAULT_KEY = 'default-encryption-key-change-in-production'.freeze

    class << self
      def encrypt(text)
        salt = SecureRandom.random_bytes(SALT_LENGTH)
        iv = SecureRandom.random_bytes(IV_LENGTH)
        cipher = OpenSSL::Cipher.new(ALGORITHM)
        cipher.encrypt
        cipher.iv_len = IV_LENGTH
        cipher.key = derive_key(salt)
        cipher.iv = iv

        encrypted = cipher.update(text) + cipher.final
        tag = cipher.auth_tag

        Base64.strict_encode64(salt + iv + tag + encrypted)
      end

      def decrypt(encrypted_text)
        buffer = Base64.decode64(encrypted_text)

        salt = buffer.byteslice(0, SALT_LENGTH)
        iv = buffer.byteslice(SALT_LENGTH, IV_LENGTH)
        tag = buffer.byteslice(SALT_LENGTH + IV_LENGTH, TAG_LENGTH)
        ciphertext = buffer.byteslice(SALT_LENGTH + IV_LENGTH + TAG_LENGTH, buffer.bytesize)

        cipher = OpenSSL::Cipher.new(ALGORITHM)
        cipher.decrypt
        cipher.iv_len = IV_LENGTH
        cipher.key = derive_key(salt)
        cipher.iv = iv
        cipher.auth_tag = tag

        cipher.update(ciphertext) + cipher.final
      end

      private

      def derive_key(salt)
        OpenSSL::PKCS5.pbkdf2_hmac(
          encryption_key,
          salt,
          ITERATIONS,
          KEY_LENGTH,
          OpenSSL::Digest::SHA256.new
        )
      end

      def encryption_key
        key = ENV['ENCRYPTION_KEY']
        key.nil? || key.empty? ? DEFAULT_KEY : key
      end
    end
  end
end
